docs: add sequence diagrams giving an overview of how to use ArmoniK RPCs and what they do internally (#876)
aneoconsulting/ArmoniK.Core: .docs/content/services/tasks_service.md

# Tasks Service RPCs actions

## Submit tasks

The following sequence diagram illustrates the internal interactions when submitting tasks in the Tasks Service:

```mermaid
sequenceDiagram
  participant c as Client
  participant cp as Control Plane
  participant s as SessionsTable
  participant t as TasksTable
  participant r as ResultsTable
  participant q as Queue

  c ->>+ cp: Submit tasks
  cp ->>+ s: Get session metadata
  s -->>- cp: #

  cp ->>+ t: Create tasks with status Creating
  t -->>- cp: #

  cp ->>+ r: Change ownership of output results
  r -->>- cp: #

>   cp ->>+ r: Get completed task input results

Reviewer: ```suggestion
  cp ->>+ r: Get completed input results
```

  r -->>- cp: #

>   cp ->>+ t: Get tasks where input results are completed

Reviewer: ```suggestion
  cp ->>+ t: Get tasks where all input results are completed
```

5. **Get Completed Task Input Results**: The Control Plane retrieves the completed task input results from the ResultsTable.
6. **Get Tasks with Completed Input Results**: The Control Plane retrieves tasks from the TasksTable where the input results are completed.
7. **Set Tasks Status to Pending**: The Control Plane updates the status of all tasks to "Pending" in the TasksTable.
8. **Insert Ready Tasks into the Queue**: The Control Plane inserts the ready tasks into the Queue.
9. **Set Queued Tasks Status to Submitted**: The Control Plane updates the status of the queued tasks to "Submitted" in the TasksTable.
10. **Response to Client**: The Control Plane sends the metadata of the created tasks back to the client.

## Cancel tasks

The following sequence diagram illustrates the internal interactions when canceling tasks in the Tasks Service:

```mermaid
sequenceDiagram
  participant c as Client
  participant cp as Control Plane
  participant t as TaskTable
  participant r as ResultTable
  participant a as Agent

  c ->>+ cp: Cancel tasks

  cp ->>+ t: Change tasks status to Cancelling
  t -->>- cp: #

  cp ->>+ t: Get cancelled tasks owner pod ids
  t -->>- cp: #

>   cp ->>+ a: Activate cancellation of tasks running on agents

Reviewer: ```suggestion
  cp ->>+ a: Request immediate cancellation of running tasks on agents
```
